File: src/dto.ts
/**
 * Data Transfer Object shapes mirrored from the Rust backend.
 *
 */
import { z } from 'zod';

export type ParsedDataTable = z.infer<typeof zParsedDataTable>;
export type AnalysisRunResult = z.infer<typeof zAnalysisRunResult>;

const zCellValue = z.union([z.string(), z.number(), z.boolean(), z.null()]);

export const zParsedDataTable = z
  .object({
    headers: z.array(z.string()),
    rows: z.array(z.array(zCellValue)),
    note: z.string().optional(),
    title: z.string().optional(),
  })
  .superRefine((val, ctx) => {
    const w = val.headers.length;
    for (let i = 0; i < val.rows.length; i++) {
      if (val.rows[i].length !== w) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `row ${i} length ${val.rows[i].length} != headers ${w}`,
          path: ['rows', i],
        });
      }
    }
  });

export const zDatasetCacheId = z.string();

export const zSheetNames = z.array(z.string());

export const zRegressionResult = z.object({
  model_summary: zParsedDataTable,
  coefficients: zParsedDataTable,
  anova: zParsedDataTable,
});

export const zFactorResult = z.object({
  eigen: zParsedDataTable,
  pattern: zParsedDataTable,
  rotmat: zParsedDataTable,
  structure: zParsedDataTable.optional(),
  phi: zParsedDataTable.optional(),
});

// 分析結果は kind 付きの判別ユニオン
export const zAnalysisTableResult = z.object({
  kind: z.literal('table'),
  table: zParsedDataTable,
});

export const zAnalysisRegressionResult = z.object({
  kind: z.literal('regression'),
  regression: zRegressionResult,
});

export const zAnalysisFactorResult = z.object({
  kind: z.literal('factor'),
  factor: zFactorResult,
});

export const zAnalysisResult = z.union([
  zAnalysisTableResult,
  zAnalysisRegressionResult,
  zAnalysisFactorResult,
]);

export type RegressionResult = z.infer<typeof zRegressionResult>;
export type FactorResult = z.infer<typeof zFactorResult>;
export type AnalysisTableResult = z.infer<typeof zAnalysisTableResult>;
export type AnalysisRegressionResult = z.infer<typeof zAnalysisRegressionResult>;
export type AnalysisFactorResult = z.infer<typeof zAnalysisFactorResult>;
export type AnalysisResult = z.infer<typeof zAnalysisResult>;

export const zAnalysisRunResult = z.object({
  result: zAnalysisResult,
  loggedAt: z.string(),
  analysisId: z.string(),
});

export const zAnalysisLogSummary = z.object({
  analysisId: z.string(),
  timestamp: z.string(),
  analysisType: z.string(),
  filePath: z.string(),
  sheetName: z.string(),
  variables: z.array(z.string()),
});

export const zAnalysisLogEntry = z.object({
  analysisId: z.string(),
  timestamp: z.string(),
  analysisType: z.string(),
  filePath: z.string(),
  sheetName: z.string(),
  variables: z.array(z.string()),
  options: z.record(z.string(), z.unknown()),
  result: zAnalysisResult,
});

export type AnalysisLogSummary = z.infer<typeof zAnalysisLogSummary>;
export type AnalysisLogEntry = z.infer<typeof zAnalysisLogEntry>;

File: src-tauri/src/dto.rs
use indexmap::IndexMap;
use serde::{
    Deserialize,
    Serialize,
};

pub type NumericDataset = IndexMap<String, Vec<Option<f64>>>;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ParsedDataTable {
    pub headers: Vec<String>,
    pub rows: Vec<Vec<serde_json::Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
}

impl ParsedDataTable {
    pub fn validate(&self) -> Result<(), String> {
        let w = self.headers.len();
        for (i, row) in self.rows.iter().enumerate() {
            if row.len() != w {
                return Err(format!("ParsedDataTable validation error: row {} length {} != headers {}",
                                   i,
                                   row.len(),
                                   w));
            }
            for (j, cell) in row.iter().enumerate() {
                match cell {
                    serde_json::Value::Null
                    | serde_json::Value::Bool(_)
                    | serde_json::Value::Number(_)
                    | serde_json::Value::String(_) => {},
                    _ => {
                        return Err(format!("ParsedDataTable validation error: rows[{}][{}] has unsupported type",
                                           i, j));
                    },
                }
            }
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct RegressionResult {
    pub model_summary: ParsedDataTable,
    pub coefficients: ParsedDataTable,
    pub anova: ParsedDataTable,
}

impl RegressionResult {
    pub fn validate(&self) -> Result<(), String> {
        self.model_summary
            .validate()
            .map_err(|e| format!("model_summary: {}", e))?;
        self.coefficients
            .validate()
            .map_err(|e| format!("coefficients: {}", e))?;
        self.anova.validate().map_err(|e| format!("anova: {}", e))?;
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct FactorResult {
    pub eigen: ParsedDataTable,
    pub pattern: ParsedDataTable,
    pub rotmat: ParsedDataTable,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub structure: Option<ParsedDataTable>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub phi: Option<ParsedDataTable>,
}

impl FactorResult {
    pub fn validate(&self) -> Result<(), String> {
        self.eigen.validate().map_err(|e| format!("eigen: {}", e))?;
        self.pattern.validate().map_err(|e| format!("pattern: {}", e))?;
        self.rotmat.validate().map_err(|e| format!("rotmat: {}", e))?;
        if let Some(structure) = self.structure.as_ref() {
            structure.validate().map_err(|e| format!("structure: {}", e))?;
        }
        if let Some(phi) = self.phi.as_ref() {
            phi.validate().map_err(|e| format!("phi: {}", e))?;
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum AnalysisResult {
    Table { table: ParsedDataTable },
    Regression { regression: RegressionResult },
    Factor { factor: FactorResult },
}

impl AnalysisResult {
    pub fn validate(&self) -> Result<(), String> {
        match self {
            AnalysisResult::Regression { regression } => regression.validate(),
            AnalysisResult::Table { table } => table.validate(),
            AnalysisResult::Factor { factor } => factor.validate(),
        }
    }
}
